linkerd/identity/src/imp/openssl.rs

```diff
@@ -0,0 +1,345 @@
+pub enum InvalidCrt {
+    SubjectAltName(String),
+    Verify(X509VerifyResult),
+    General(Error),
+}
+
+impl From<Error> for InvalidCrt {
+    fn from(err: Error) -> Self {
+        InvalidCrt::General(err)
+    }
+}
+
+impl From<ErrorStack> for InvalidCrt {
+    fn from(err: ErrorStack) -> Self {
+        InvalidCrt::General(err.into())
+    }
+}
+
+impl fmt::Display for InvalidCrt {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            InvalidCrt::SubjectAltName(name) => write!(f, "Subject alt name incorrect {}", name),
+            InvalidCrt::Verify(err) => err.fmt(f),
+            InvalidCrt::General(err) => err.fmt(f),
+        }
+    }
+}
+
+impl error::Error for InvalidCrt {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            InvalidCrt::Verify(err) => err.source(),
+            InvalidCrt::General(err) => err.source(),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CrtKey {
+    id: LocalId,
+    expiry: SystemTime,
+    client_config: Arc<ClientConfig>,
+    server_config: Arc<ServerConfig>,
+}
+
+// === CrtKey ===
+impl CrtKey {
+    pub fn name(&self) -> &Name {
+        self.id.as_ref()
+    }
+
+    pub fn expiry(&self) -> SystemTime {
+        self.expiry
+    }
+
+    pub fn id(&self) -> &LocalId {
+        &self.id
+    }
+
+    pub fn client_config(&self) -> Arc<ClientConfig> {
+        self.client_config.clone()
+    }
+
+    pub fn server_config(&self) -> Arc<ServerConfig> {
+        self.server_config.clone()
+    }
+}
+
+impl fmt::Debug for CrtKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.debug_struct("CrtKey")
+            .field("id", &self.id)
+            .field("expiry", &self.expiry)
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Crt {
+    pub(crate) id: LocalId,
+    expiry: SystemTime,
+    pub cert: X509,
+    pub chain: Vec<X509>,
+}
+
+impl Crt {
+    pub fn new(
+        id: LocalId,
+        leaf: Vec<u8>,
+        intermediates: Vec<Vec<u8>>,
+        expiry: SystemTime,
+    ) -> Self {
+        let mut chain = Vec::with_capacity(intermediates.len() + 1);
+        let cert = X509::from_der(&leaf).unwrap();
+        chain.extend(
+            intermediates
+                .into_iter()
+                .map(|crt| X509::from_der(&crt).unwrap()),
+        );
+
+        Self {
+            id,
+            cert,
+            chain,
+            expiry,
+        }
+    }
+
+    pub fn name(&self) -> &Name {
+        self.id.as_ref()
+    }
+}
+
+#[derive(Clone)]
+pub struct ClientConfig {
+    pub root_certs: Arc<X509Store>,
+    pub key: Option<Arc<Key>>,
+    pub cert: Option<Arc<Crt>>,
+    pub protocols: Arc<Vec<Vec<u8>>>,
+}
+
+impl fmt::Debug for ClientConfig {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&self.protocols, f)
+    }
+}
+
+impl ClientConfig {
+    pub fn new(
```

Comment:
it looks like the `new` constructor is never called outside this module, can we just change this to
```suggestion
    fn new(
```
since this constructor exposes an implementation detail (the `X509Store` type from `openssl`).

Reply:
Absolutely.